[JSC] TypedArray change-array-by-copy should be C++

Constellation opened the pull request:
#### 9bbad8a7d16c16aa00440cb190b310e46a4ec5e2
<pre>
[JSC] TypedArray change-array-by-copy should be C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=243162">https://bugs.webkit.org/show_bug.cgi?id=243162</a>

Reviewed by Alexey Shvayka and Ross Kirsling.

This patch implements TypedArray change-array-by-copy in C++ to get performance boost.
TypedArray has a lot of characteristics of non-dynamic behavior. Since it is typed too,
then we can easily write super fast C++ implementations for these methods.
This patch keeps toSorted in JS while large part of it is implemented in C++ already.
And we also migrate subarray from JS to C++ since we have fast speciesConstruct now.

                                            ToT                     Patched

    typed-array-to-sorted-hot        131.0737+-0.0586     ^     79.5521+-0.1355        ^ definitely 1.6476x faster
    typed-array-subarray-hot           4.5564+-0.2404     ^      3.5946+-0.0116        ^ definitely 1.2676x faster
    typed-array-to-reversed-hot       93.7245+-0.1785     ^     28.5393+-0.1813        ^ definitely 3.2841x faster
    typed-array-with-hot              78.6775+-0.2573     ^     10.2874+-0.0542        ^ definitely 7.6479x faster

* JSTests/microbenchmarks/typed-array-subarray-hot.js: Added.
* JSTests/microbenchmarks/typed-array-to-reversed-hot.js: Added.
* JSTests/microbenchmarks/typed-array-to-sorted-hot.js: Added.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/TypedArrayPrototype.js:
(subarray): Deleted.
(toReversed): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::speciesConstruct):
(JSC::genericTypedArrayViewProtoFuncToReversed):
(JSC::genericTypedArrayViewProtoFuncToSorted):
(JSC::genericTypedArrayViewProtoFuncSlice):
(JSC::genericTypedArrayViewProtoFuncSubarray):
(JSC::genericTypedArrayViewPrivateFuncSubarrayCreate): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.h:

Canonical link: <a href="https://commits.webkit.org/252891@main">https://commits.webkit.org/252891@main</a>
</pre>

shvaikalesh commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -718,17 +774,13 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSubarrayCreate(VM&
-    ASSERT(callFrame->argument(1).isUndefined() || callFrame->argument(1).isNumber());
     size_t begin = argumentClampedIndexFromStartOrEnd(globalObject, callFrame->argument(0), thisLength);
-    scope.assertNoException();
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
nit:
```suggestion
    RETURN_IF_EXCEPTION(scope, { });
```

Constellation replied:
Fixed.

shvaikalesh commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -718,17 +774,13 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSubarrayCreate(VM&
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
     size_t end = argumentClampedIndexFromStartOrEnd(globalObject, callFrame->argument(1), thisLength, thisLength);
-    scope.assertNoException();
+    RETURN_IF_EXCEPTION(scope, encodedJSValue());
```
nit:
```suggestion
    RETURN_IF_EXCEPTION(scope, { });
```

Constellation replied:
Fixed.

shvaikalesh commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -746,32 +798,94 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSubarrayCreate(VM&
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+
+    if (UNLIKELY(thisObject->isDetached())) {
```
Ah, I wish one day we would change the spec's [`IsValidIntegerIndex`](https://tc39.es/ecma262/#sec-isvalidintegerindex) not to include detached buffer check.

rkirsling replied:
I'm not sure there's any spec work left to be done there though; we stopped various operations from throwing, but given that the concept of a detached buffer exists, we do need to early out before attempting to interact with its contents.

The only redundancy I see is that integer-indexed [`[[DefineOwnProperty]]`](https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects-defineownproperty-p-desc) ends up checking `IsValidIntegerIndex` twice because of the `IntegerIndexedElementSet` call.
(Edit: Oops, and even that's not actually redundant, because ToNumber/ToBigInt of the value to set could detach the buffer.)

shvaikalesh commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -746,32 +798,94 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSubarrayCreate(VM&
+        throwVMRangeError(globalObject, scope, "index should not be negative zero"_s);
+        return;
+    }
+    if (UNLIKELY(index < 0 || index >= thisObject->length())) {
```
nit:
```suggestion
    if (UNLIKELY(!thisObject->inBounds(index))) {
```

Constellation replied:
Nice. Fixed.

Constellation replied:
Ah, we need to have index < 0 check still.

shvaikalesh reviewed (approved):
r=me, absolutely stunning!

rkirsling commented on Source/JavaScriptCore/builtins/TypedArrayPrototype.js:
```diff
@@ -393,85 +375,28 @@ function at(index)
-    else
-        actualIndex = length + relativeIndex;
+    if (comparator !== @undefined && !@isCallable(comparator))
+        @throwTypeError("TypedArray.prototype.toSorted requires the comparator argument to be a function or undefined");
```
Looks like ArrayPrototype.js calls the argument `comparefn` (following spec naming) but says "comparator" in the error message (which is certainly more pleasant). Should we keep this in sync?

Constellation replied:
Yeah, we should use `comparator`.

rkirsling commented on Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
```diff
@@ -746,32 +798,94 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSubarrayCreate(VM&
+    typename ViewClass::ElementType nativeValue = ViewClass::toAdaptorNativeFromValue(globalObject, callFrame->argument(1));
     RETURN_IF_EXCEPTION(scope, { });
 
-    JSArrayBufferView* validated = validateTypedArray(globalObject, result);
```
After this, the only callsite left for `validateTypedArray` is in AtomicsObject, so that might warrant some cleanup.

rkirsling reviewed (approved):
r=me as well, though given the title, I was surprised that `toSorted` is remaining in JS while the existing `subarray` method is moving to C++. Can you make this explicit in the changelog?

Constellation committed:
[JSC] TypedArray change-array-by-copy should be C++
https://bugs.webkit.org/show_bug.cgi?id=243162

Reviewed by Alexey Shvayka and Ross Kirsling.

This patch implements TypedArray change-array-by-copy in C++ to get performance boost.
TypedArray has a lot of characteristics of non-dynamic behavior. Since it is typed too,
then we can easily write super fast C++ implementations for these methods.
This patch keeps toSorted in JS while large part of it is implemented in C++ already.
And we also migrate subarray from JS to C++ since we have fast speciesConstruct now.

                                            ToT                     Patched

    typed-array-to-sorted-hot        131.0737+-0.0586     ^     79.5521+-0.1355        ^ definitely 1.6476x faster
    typed-array-subarray-hot           4.5564+-0.2404     ^      3.5946+-0.0116        ^ definitely 1.2676x faster
    typed-array-to-reversed-hot       93.7245+-0.1785     ^     28.5393+-0.1813        ^ definitely 3.2841x faster
    typed-array-with-hot              78.6775+-0.2573     ^     10.2874+-0.0542        ^ definitely 7.6479x faster

* JSTests/microbenchmarks/typed-array-subarray-hot.js: Added.
* JSTests/microbenchmarks/typed-array-to-reversed-hot.js: Added.
* JSTests/microbenchmarks/typed-array-to-sorted-hot.js: Added.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/TypedArrayPrototype.js:
(subarray): Deleted.
(toReversed): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::speciesConstruct):
(JSC::genericTypedArrayViewProtoFuncToReversed):
(JSC::genericTypedArrayViewProtoFuncToSorted):
(JSC::genericTypedArrayViewProtoFuncSlice):
(JSC::genericTypedArrayViewProtoFuncSubarray):
(JSC::genericTypedArrayViewPrivateFuncSubarrayCreate): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.h:

Canonical link: https://commits.webkit.org/252891@main
